src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/24.diff.sql

```diff
@@ -0,0 +1,152 @@
+        ResourceId,
+        ResourceTypeId,
+        ResourceVersion,
+        ResourceChangeTypeId
+    FROM PartitionBoundaries AS p
+    CROSS APPLY (
+        /* Given the fact that Read Committed Snapshot isolation level is enabled on the FHIR database,
+           using TABLOCK and HOLDLOCK table hints to avoid skipping resource changes 
+           due to interleaved transactions on the resource change data table. */
+        SELECT TOP(@pageSize) Id,
+            Timestamp,
+            ResourceId,
+            ResourceTypeId,
+            ResourceVersion,
+            ResourceChangeTypeId
+        /* Acquires and holds the table lock to prevent new resource changes from being created during the select query execution.
+           Without the TABLOCK and HOLDLOCK hints, the lock will be applied at the row or page level which can result in missed reads
+           if an insert occurs in range that was previously read and the locks were released on the read. */
+        FROM dbo.ResourceChangeData WITH (TABLOCK, HOLDLOCK)
+            WHERE Id >= @startId
+                AND $PARTITION.PartitionFunction_ResourceChangeData_Timestamp(Timestamp) = $PARTITION.PartitionFunction_ResourceChangeData_Timestamp(p.PartitionBoundary)
+        ORDER BY Id ASC
+        ) AS rcd
+    ORDER BY rcd.Id ASC;
+END;
+GO
+
+IF EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'PK_ResourceChangeData_TimestampId')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Deleting PK_ResourceChangeData_TimestampId index from ResourceChangeData table.';
+
+    /* Drops PK_ResourceChangeData_TimestampId index. "ONLINE = ON" indicates long-term table locks aren't held for the duration of the index operation. 
+       During the main phase of the index operation, only an Intent Share (IS) lock is held on the source table. 
+       This behavior enables queries or updates to the underlying table and indexes to continue. */
+    ALTER TABLE dbo.ResourceChangeData DROP CONSTRAINT PK_ResourceChangeData_TimestampId WITH(ONLINE = ON);
+END;
+GO
+
+IF EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'PK_ResourceChangeDataStaging_TimestampId')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Deleting PK_ResourceChangeDataStaging_TimestampId index from ResourceChangeDataStaging table.';
+
+    /* Drops PK_ResourceChangeDataStaging_TimestampId index. "ONLINE = ON" indicates long-term table locks aren't held for the duration of the index operation. 
+       During the main phase of the index operation, only an Intent Share (IS) lock is held on the source table. 
+       This behavior enables queries or updates to the underlying table and indexes to continue. */
+    ALTER TABLE dbo.ResourceChangeDataStaging DROP CONSTRAINT PK_ResourceChangeDataStaging_TimestampId WITH(ONLINE = ON);
+END;
+GO
+
+IF NOT EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'IXC_ResourceChangeData')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Creating IXC_ResourceChangeData index on ResourceChangeData table.';
```

Comment:
Might be this can move upward just after dropping the primary key to ensure that table is not left without index for longer time

Reply:
A reason we put the order in a current way is that:
- online index operation is used.
- partition switch-out operation requires that the main table and staging table have exactly the same structure and indexes. 
Considering these, we would like to keep the two table structures and indexes the same if migration is failed right after dropping the index from a staging table or right before creating an index on the main table.
  
I would like to highlight that dropping an index from a staging table will be very fast due staging table is always empty, except during partition switch-out operation.